### PR TITLE
Support cards in status views

### LINKED ISSUE #234: Proposal: Preview to message links.
Propose new feature: In case a message contains a link to a website, show preview of that website.

### LINKED ISSUE #518: [FEATURE REQUEST] Render display cards for links in app timeline
Most social apps have the ability to pre-fetch a URL and render a display card from markdown/metadata. This provides a richer experience as it allows a user to understand what is at the link and provides better visual cues to the information being linked to.

Twitter has [documented its implementation quite well here](https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/abouts-cards). 

It seems Mastodon web client sorta renders it, but not very well. The IOS app doesn't even attempt to render a card.

My request would be to follow the same formatting as other social apps. I know for [my blog](https://danaepp.com), they all render out the card for my posts quite well, but differently. ie, Some render extra metadata, while others do not. Twitter will show the card image, title, and description... where LinkedIn also renders how long the read would be for the article.

I'll link a few examples to show the difference.  

- [Twitter card](https://infosec.exchange/system/media_attachments/files/109/298/495/238/334/202/original/a7c968182757c17c.jpg)
- [LinkedIn card](https://infosec.exchange/system/media_attachments/files/109/298/490/008/285/547/original/95483e987fcc971a.jpg)

### PR DESCRIPTION
**Pull request by kylebshr**
Fixes https://github.com/mastodon/mastodon-ios/issues/234
Fixes https://github.com/mastodon/mastodon-ios/issues/518

This PR persists cards to core data, and displays them in the feed using a layout fairly similar to Twitter.

Examples:

|  | Regular | Scaled |
|-|-|-|
| Compact | Simulator Screen Shot - iPhone 14 Pro - 2022-11-26 at 21 44 53 | Simulator Screen Shot - iPhone 14 Pro - 2022-11-26 at 21 45 16 |
| Large | Simulator Screen Shot - iPhone 14 Pro - 2022-11-26 at 21 45 32 | Simulator Screen Shot - iPhone 14 Pro - 2022-11-26 at 21 45 23 |

Inline:

| Compact | Large |
|-|-|
| Simulator Screen Shot - iPhone 14 Pro - 2022-11-26 at 21 47 18 | Simulator Screen Shot - iPhone 14 Pro - 2022-11-26 at 21 46 00 |

TODO:

- [x] ~Make a ViewModel for the `LinkPreviewButton` rather than using `Card` directly~ Don't think this is really necessary since it's static
- [x] Rename `LinkPreviewButton` → `PreviewCardButton` or something
- [x] Fix layout bug with large card where image is incorrectly positions when first scrolled into view
- [x] Remove debug prints
- [x] Refine layout

### CODE REVIEW

File: MastodonSDK/Sources/MastodonUI/View/Content/LinkPreviewButton.swift [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        clipsToBounds = true
+        layer.cornerCurve = .continuous
+        layer.cornerRadius = 10
+
+        highlightView.backgroundColor = UIColor.black.withAlphaComponent(0.1)
+        highlightView.isHidden = true
+
+        titleLabel.numberOfLines = 2
+        titleLabel.textColor = Asset.Colors.Label.primary.color
+        titleLabel.font = .preferredFont(forTextStyle: .body)
```
Comment (j-f1) [inline MastodonSDK/Sources/MastodonUI/View/Content/LinkPreviewButton.swift]: Heads up that you may need to reapply all the fonts in a `traitCollectionDidChange` method to pick up font size settings while the app is running. (I haven’t tested so `preferredFont` may auto-scale itself)

Comment (kylebshr) [reply MastodonSDK/Sources/MastodonUI/View/Content/LinkPreviewButton.swift]: `preferredFont` auto scales! There are screenshots in the PR now.

**Comment by kylebshr** [comment]
I've been banging my head against the wall trying to figure out why the large layout sometimes lays out the image 4.66pts too small and 2.33pts offset (you can see this in forms.swift.org card). I'm gonna say this is good to go. If anyone else wants to try to debug, please do.

File: MastodonSDK/Sources/MastodonUI/View/Content/StatusCardControl.swift [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        NSLayoutConstraint.activate(layoutConstraints)
+    }
+
+    private func icon(for layout: Layout) -> UIImage? {
+        switch layout {
+        case .compact:
+            return UIImage(systemName: "newspaper.fill")
+        case .large:
+            let configuration = UIImage.SymbolConfiguration(pointSize: 32)
+            return UIImage(systemName: "photo", withConfiguration: configuration)
```
Comment (j-f1) [inline MastodonSDK/Sources/MastodonUI/View/Content/StatusCardControl.swift:168]: Under what circumstances would this be used? one of the things I dislike about Twitter cards is when the link has no image but still gets the large image layout. Would it be possible to switch back to the other layout if there’s no image?

Comment (kylebshr) [reply MastodonSDK/Sources/MastodonUI/View/Content/StatusCardControl.swift:168]: If there's no image URL, it'll use the compact layout. This is used as a placeholder and backup if the image fails to load.

**Comment by kylebshr** [comment]
> I've been banging my head against the wall trying to figure out why the large layout sometimes lays out the image 4.66pts too small and 2.33pts offset (you can see this in forms.swift.org card). I'm gonna say this is good to go. If anyone else wants to try to debug, please do.

I think this is resolved?

File: MastodonSDK/Sources/MastodonUI/View/Content/StatusCardControl.swift [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        case large(aspectRatio: CGFloat)
+    }
+}
+
+private extension Card {
+    var layout: StatusCardControl.Layout {
+        return width == height || image == nil
+        ? .compact
+        : .large(aspectRatio: CGFloat(width) / CGFloat(height))
+    }
```
Comment (j-f1) [inline MastodonSDK/Sources/MastodonUI/View/Content/StatusCardControl.swift:190]: I saw a forums.swift.org link with a 397×400 image. This change will treat slightly non-square images as square too.

```suggestion
    var layout: StatusCardControl.Layout {
        let aspectRatio = CGFloat(width) / CGFloat(height)
        return abs(aspectRatio - 1) < 0.05 || image == nil
        ? .compact
        : .large(aspectRatio: aspectRatio)
    }
```

**Comment by j-f1** [comment]
I’m also curious what you think about putting divider line between the image and the text. It feels a bit weird to me when the image has a white background, but what do you think?

**Comment by j-f1** [comment]
Also: it's a bit weird to me that the background of the card darkens more than the table cell as a whole when you tap on the content of the post. IMO it should only change color if you tap directly on it.

**Comment by kylebshr** [comment]
Feel free to make any of those changes; I don't think I'll have time to work on this any more. Anyone is welcome to take over the work.

**Pull request closed without merging**